webapi/src/main/scala/org/knora/webapi/slice/export/api/ExportRestService.scala

```diff
@@ -29,27 +30,42 @@ final case class ExportRestService(
     user: User,
   )(
     request: ExportRequest,
-  ): IO[V3ErrorInfo, (String, MediaType, String)] =
+  ): IO[V3ErrorInfo, (String, MediaType, String)] = {
     (for {
-      resourceClassIri <- iriConverter.asResourceClassIri(request.resourceClass)
-      shortcode        <- ZIO.fromEither(resourceClassIri.smartIri.getProjectShortcode)
-      properties       <- ZIO.foreach(request.selectedProperties)(iriConverter.asPropertyIri)
+      resourceClassIri <- iriConverter.asResourceClassIri(request.resourceClass).mapError(BadRequest(_))
+      shortcode        <- ZIO.fromEither(resourceClassIri.smartIri.getProjectShortcode).mapError(BadRequest(_))
+      properties       <- ZIO.foreach(request.selectedProperties)(iriConverter.asPropertyIri).mapError(BadRequest(_))
 
+      project    <- projectService.findByShortcode(shortcode).errorOrNotFound(NotFound(resourceClassIri))
       ontologyIri = resourceClassIri.ontologyIri
-      project    <- projectService.findByShortcode(shortcode).orDie.someOrFail(NotFound.from(shortcode))
-      _          <- ontologyService.findById(ontologyIri).someOrFail(NotFound.notfound(ontologyIri))
+      _          <- ontologyService.findById(ontologyIri).someOrFail(NotFound(ontologyIri))
 
-      data           <- exportService.exportResources(project, resourceClassIri, properties, user, request.language).orDie
+      data <-
+        exportService
+          .exportResources(project, resourceClassIri, properties, user, request.language, request.includeResourceIri)
+          .orDie
       (headers, rows) = data
       csv            <- ZIO.scoped(csvService.writeToString(rows)(using ExportedResource.rowBuilder(headers))).orDie
       now            <- Clock.instant
     } yield (
       csv,
       MediaType.TextCsv,
       s"attachment; filename=project_${shortcode.value}_resources_${resourceClassIri.name}_${now}.csv",
-    )).mapError(t => BadRequest(t.toString))
+    ))
+  }.flatMapError {
```

Comment:
The problem with the `flatMapError` is that when you read it you do not know where the `Throwable` is coming from. The error type of the for comprehension should rather be `V3ErrorInfo` directly.

I am not a big fan of this "catch all" type of error handling. 

I think it would be better to `.orDie` on the `projectService.findByShortcode` and `ontologyService.findById` calls.